modules/git/utils_closererror.go

```diff
@@ -0,0 +1,99 @@
+// Copyright 2022 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package git
+
+import (
+	"io"
+	"sync"
+)
+
+type closerError struct {
+	io.Closer
+	err  error
+	lock sync.Mutex
+}
+
+func (c *closerError) CloseWithError(err error) error {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	if c.err != nil {
+		return c.err
+	}
+	if c.Closer != nil {
+		_ = c.Closer.Close()
+	}
+	c.err = err
```

Comment:
To be more in-line with current implementation:
```suggestion
	if err == nil {
		err = io.ErrClosedPipe
	}
	c.err = err	
```